beacon-chain/state/state-native/setters_withdrawal.go

```diff
@@ -100,3 +100,24 @@ func (b *BeaconState) DequeuePendingPartialWithdrawals(n uint64) error {
 
 	return nil
 }
+
+// SetPendingPartialWithdrawals sets the pending partial withdrawals. This method mutates the state.
+func (b *BeaconState) SetPendingPartialWithdrawals(pendingPartialWithdrawals []*eth.PendingPartialWithdrawal) error {
+	if b.version < version.Electra {
+		return errNotSupported("SetPendingPartialWithdrawals", b.version)
+	}
+
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	if pendingPartialWithdrawals == nil {
+		return errors.New("cannot set nil pending partial withdrawals")
+	}
+	b.sharedFieldReferences[types.PendingPartialWithdrawals].MinusRef()
+	b.sharedFieldReferences[types.PendingPartialWithdrawals] = stateutil.NewRef(1)
+
+	b.pendingPartialWithdrawals = pendingPartialWithdrawals
+	b.markFieldAsDirty(types.PendingPartialWithdrawals)
```

Comment:
why do we not need to rebuild trie here for setting it
like we do with b.rebuildTrie[types.PendingPartialWithdrawals] = true

Reply:
Field tries is something I never understood clearly, even though Nishant explained them to me a few times... That being said, they are used when computing roots of fields and for pending partial withdrawals, we don't rely on a trie to compute the root. See https://github.com/OffchainLabs/prysm/blob/cc2565a422f7db0784f63883d561f9b606995911/beacon-chain/state/stateutil/pending_partial_withdrawals_root.go#L9

Reply:
When we make changes to a beacon state field, those changes affect not only the field itself, but all the hashes up to the hash tree root of the full state. In order to avoid hashing the entire state over and over again, we keep the full hash tree root of the state in memory. However, we do not want to update it on every field modification, because modifying for example two consecutive balances would most likely render an update after the first modification useless (all of the hashes will most likely need to be recomputed).

Therefore, when we update a field, we mark the field as dirty to remember that we need to compute that path. When we change large sections though (typically large slices), we immediately call to rebuild the trie because of this.